src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs

```diff
@@ -55,7 +55,9 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             foreach (var diagnostic in diagnostics)
             {
-                var localDeclaration = (LocalDeclarationStatementSyntax)diagnostic.AdditionalLocations[0].FindNode(cancellationToken);
+                var declarationNode = diagnostic.AdditionalLocations[0].FindNode(cancellationToken);
+                var localDeclaration = declarationNode as LocalDeclarationStatementSyntax
+                                    ?? (LocalDeclarationStatementSyntax)declarationNode.Parent;
```

Comment:
Can we make this more conservative, say:

```
var localDeclaration = declarationNode as LocalDeclarationStatementSyntax ?? declarationNode.Parent as LocalDeclarationStatementSyntax;
if (localDeclaration == null)
{
  return;
}
```

Reply:
note: this is fix-all. So if we get null here, we likely shoudl just continue, not return 'null'.

Reply:
Also, i would prefer we not take this fix like this.  Instead of having the fixer side bailing out.  We should simply not be registering the diagnostic if we can't fix this.  so this shuld be on hte analyzer side.  if our AdditionalLocations[0] is not a LocalDeclStatement there we likely shouldn't register in the first palce.

Reply:
Looking at the analyzer, this case seems even more perplexing. It seems the first additional location is indeed a local declaration: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs,130

Reply:
Indeed.  I don't understand this PR then :)

Reply:
@CyrusNajmabadi @mavasani the call to `FindNode` is finding the `VariableDeclarationSyntax` instead of the `LocalDeclarationStatementSyntax`. I _think_ that's ok, and not an issue with `FindNode`. Given that behavior, maybe this should be 

```C#
var localDeclaration = declarationNode is LocalDeclarationStatementSyntax ? declarationNode : (LocalDeclarationStatementSyntax)declarationNode.Parent; 
```

Thoughts?

Reply:
> @CyrusNajmabadi @mavasani the call to FindNode is finding the VariableDeclarationSyntax

Why is it finding that though? Based on http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs,130 then `AdditionalLocations[0]` should always be a LocalDeclarationStatementSyntax.  It's unclear to me under what circumstances that would ever find a VariableDeclarationNode.

Reply:
Maybe an error in how the syntax is parsed? The example I have is from #29793 

```C#
Func<Task> f_irstAction = () =>
            {
                return instance.ExecuteUnderLockAsync(async (ct) =>
                {
                    firstEntered.Set();
                    await firstRelease;
                }, CancellationToken.None);
            });
```

The trailing `)` is included as part of the `LocalDeclarationStatementSyntax` span but not the `VariableDeclarationSyntax` span. With that difference, the check in FindNode doesn't keep going up the tree for evaluation [http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,829](http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,829)

Reply:
Note that this is invalid code and won't compile, which is called out in the issue.

Reply:
Yeah, it def sounds like something is off.  SyntaxNode.GetLocation().FindNode should roundtrip *except* for cases where two nodes have the *exact same* full span (i.e. an Expression inside an Argument).  And, in that case, that's what `getInnermostNodeForTie` is for.

Reply:
> If the code location used FullSpan it would resolve correctly

`Location` needs to use 'Span` since it's definitely intended to be the right place to 'navigate to' or to 'highlight/squiggle'.  So, it sounds like FindNode should probably handle this sort of situation better.

Reply:
Hi, 

Sorry for not responding.

@ryzngard is right, the problem is that location of LocalDeclaration is parsed as VariableDeclaration in that case.

Reply:
> @ryzngard is right, the problem is that location of LocalDeclaration is parsed as VariableDeclaration in that case.

I don't know what it means when you say "location ... is parsed as ...".  Locations are not parsed.  They're just a Tree+Span.  Since hte Span originally came from a LocalDeclarationStatement, then 'Finding' that Location should round-trip and return that LocalDeclarationStatement again, not a VariableDeclaration.

Reply:
@CyrusNajmabadi I agree, `GetLocation().FindNode()` should round trip correctly, but in this case it's not. I'm not sure what the best coarse of action is here. Is a fix in the `FindNode` logic needed? Is there a better way to create and hand off the location from the analyzer side?

Reply:
>  Is a fix in the FindNode logic needed?

That would be ideal to me.  

> Is there a better way to create and hand off the location from the analyzer side?


Not that i can think of. This is the idiomatic way we recommend doing things.

Reply:
@lkts would you like to take over that change? If not, I'm happy to put up a PR and drive changes to SyntaxNode. After digging, I think FindNode logic makes sense and location generation should be changed. [This commit](https://github.com/ryzngard/roslyn/commit/8838a6252479adc26cb6669d57649530df8c2292) is roughly what I think should be changed.

Reply:
@ryzngard I do not plan to work on it now. I will close the PR.

Thank you.